Implement `BYDAY` with offset and limiting behavior

axunonb opened the pull request:
The following `RRULE` cases were not implemented (see the diff for newly added unit tests for details):

1. YEARLY + BYMONTH + numeric BYDAY offsets
   - Pattern: `FREQ=YEARLY;BYMONTH=6,9;BYDAY=2MO`
   - Semantics: treat numeric BYDAY as “nth weekday inside each BYMONTH” (produce the 2nd Monday of each specified month).

2. YEARLY + numeric BYDAY without BYMONTH
   - Pattern: `FREQ=YEARLY;BYDAY=20MO`
   - Semantics: interpret numeric BYDAY as “nth weekday of the YEAR” (the 20th Monday of the year).

3. YEARLY + BYMONTH + negative numeric BYDAY
   - Pattern: `FREQ=YEARLY;BYMONTH=6,9;BYDAY=-1SU`
   - Semantics: negative offsets select from the end of the month (e.g. `-1SU` → last Sunday in each BYMONTH).

4. Implemented the per-month offset expansion and preserved BYWEEKNO / BYMONTH compatibility and negative-offset

These behaviors are closely related and required special handling in `RecurrencePatternEvaluator.GetAbsWeekDays` and `GetAbsWeekDaysYearlyPerMonthOffsets`.

[Edit]
Relevant unit tests in `RecurrenceTestCases.txt` contain a disclaimer regarding RFC5545 interpretation in ical.net

Fixes #782
Closes #888

axunonb committed:
Implement `BYDAY` with offset and limiting behavior

The following `RRULE` cases were not implemented (see the diff for newly added unit tests for details):

1. YEARLY + BYMONTH + numeric BYDAY offsets
   - Pattern: `FREQ=YEARLY;BYMONTH=6,9;BYDAY=2MO`
   - Semantics: treat numeric BYDAY as “nth weekday inside each BYMONTH” (produce the 2nd Monday of each specified month).

2. YEARLY + numeric BYDAY without BYMONTH
   - Pattern: `FREQ=YEARLY;BYDAY=20MO`
   - Semantics: interpret numeric BYDAY as “nth weekday of the YEAR” (the 20th Monday of the year).

3. YEARLY + BYMONTH + negative numeric BYDAY
   - Pattern: `FREQ=YEARLY;BYMONTH=6,9;BYDAY=-1SU`
   - Semantics: negative offsets select from the end of the month (e.g. `-1SU` → last Sunday in each BYMONTH).

4. Implemented the per-month offset expansion and preserved BYWEEKNO / BYMONTH compatibility and negative-offset

These behaviors are closely related and required special handling in `RecurrencePatternEvaluator.GetAbsWeekDays` and `GetAbsWeekDaysYearlyPerMonthOffsets`.

Fixes #782

axunonb commented on Ical.Net.Tests/RecurrenceTests.cs:
```diff
@@ -4807,4 +4806,132 @@ public void GetOccurrences_WithMultipleOverridesForSameRecurrenceId_ShouldUseLat
+        var expected = new[]
+        {
+            new Period(new CalDateTime(2026, 6, 1), Duration.FromDays(1)),
+            new Period(new CalDateTime(2027, 6, 8), Duration.FromDays(1)),
```
https://github.com/ical-org/ical.net/issues/782#issuecomment-2853505427:
>  **The month seems to be fixed to June because BYMONTH is not present, and BYDAY is only limiting.**
>  To apply the limit, the actual dates must be determined first, which requires the month, so it must be inferred from 
>  However, there is some room for alternative interpretations. I.e., one could argue that the day is fully specified
>  via `BYDAY`, making it unnecessary to derive the month from DTSTART. Since BYDAY is limiting, we assume the month is taken from DTSTART.

Following the the first sentence: No. The current implementation has BYDAY **only limiting**, and it can't "escape".

axunonb committed:
Unit tests: Introduce OccurrenceTester as a helper usable for all test classes

maknapp commented:
FYI, I made several tests related to this for a new recurrence evaluator. I ran those tests against this fix and they all pass.

axunonb committed:
Replace CalCalc helpers with hardcoded dates in tests

Updated recurrence yearly offset tests to use hardcoded CalDateTime values instead of CalCalc helper methods. Removed CalCalc.cs as its methods are no longer referenced.

axunonb committed:
Move `RecurrenceYearlyWithOffsetsTests` to `RecurrenceTestCases.txt`

Move tests are for RFC errata 1913 (https://www.rfc-editor.org/errata_search.php?rfc=1913&amp;eid=1913)
and contain the disclaimer "Other iCalendar libraries may interpret this differently"

minichma commented on Ical.Net.Tests/Calendars/Recurrence/RecurrenceTestCases.txt:
```diff
@@ -181,3 +181,94 @@ INSTANCES:20250101,20251231,20260101
+# 
+# These 12 tests verify the behavior with the interpretation that
+# * when only BYMONTH is present, the numeric BYDAY offset applies within the month,
+# * when BYWEEKNO is present, the numeric BYDAY offset applies within the week.
```
Applying the numeric offset within the week is a reasonable interpretation following the logic of the case that `BYMONTH` is present. However, as the RFC is quite clear that BYWEEKNO + numeric BYDAY offset is not allowed, I'd rather suggest to throw in this case. Or are there cases in the wild that force us to deal with this case?

Moreover, as 1 would be the only reasonable integer (even -1 wouldn't be allowed, as every week always has 7 days, so there is no need for -1 as it isn't needed and therefore not allowed for other components like hours, which don't vary in different intervals) there isn't much point in using it, so any use of it would most likely be a usage error anyways.

axunonb replied:
Will add "RFC says that BYWEEKNO + numeric BYDAY is invalid. We're taking the lenient approach here."

minichma commented on Ical.Net.Tests/Calendars/Recurrence/RecurrenceTestCases.txt:
```diff
@@ -181,3 +181,94 @@ INSTANCES:20250101,20251231,20260101
+EXCEPTION-STEP:Enumeration
+
+# Yearly BYWEEKNO=10 BYDAY=1MO - valid: Monday of ISO week 10 each year
+RRULE:FREQ=YEARLY;BYWEEKNO=10;BYDAY=1MO;COUNT=3
```
As outlined above, maybe rather throw in this case?

minichma commented on Ical.Net/Evaluation/RecurrencePatternEvaluator.cs:
```diff
@@ -636,15 +638,25 @@ private static IEnumerable<CalDateTime> GetDayVariants(IEnumerable<CalDateTime>
+                // When limiting with an offset (e.g. "22MO" or "1MO"), compute the
+                // absolute dates for that WeekDay in the appropriate scope and
+                // check if the candidate matches one of them.
+                return GetAbsWeekDays(date, weekDay, pattern).Any(d => d.Equals(date));
```
We have the same kind of code in the expand variant, both can be pretty expensive. Also the implementations of `GetAbsWeekDays*` could be significantly improved. Candidates for the future. :-)

minichma reviewed (approved):
Very nice! Sorry for the delay!

axunonb committed:
Implement recommendations from review

axunonb commented:
@minichma, @maknapp Thanks for your reviews, and happy holidays.
